opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/plugin/LogCorrelation.java

```diff
@@ -75,9 +78,16 @@ public LogCorrelation withBaggage(final Map<String, String> mapping) {
 
     @Override
     public void onActivated(final Scope scope, final Span span) {
+        if (existingContextStack.get() == null)
+            existingContextStack.set(new Stack<>());
+        Map<String, String> contextMap = new HashMap<>();
         seeds.forEach((seed, contextKey) ->
-                Optional.ofNullable(seed.valueOf(span)).ifPresent(value ->
-                        MDC.put(contextKey, value)));
+                Optional.ofNullable(seed.valueOf(span)).ifPresent(value -> {
+                    Optional.ofNullable(MDC.get(contextKey)).ifPresent(x -> contextMap.put(contextKey, x));
```

Comment:
Preserving existing values while iterating the seeds like this is problematic.
1. It doesn't handle baggage context keys correctly.
2. It over-complicates the loop, i.e. it's hard to understand now.

I'd go with something like:

```java
    @Override
    public void onActivated(final Scope scope, final Span span) {
        contexts.get().push(preserveContext());

        seeds.forEach((seed, contextKey) ->
            Optional.ofNullable(seed.valueOf(span))
                    .ifPresent(value -> MDC.put(contextKey, value)));
    }

    private Map<String, String> preserveContext() {
        final Map<String, String> context =
                newHashMapWithExpectedSize(seeds.size() + baggage.size());

        Stream.concat(seeds.values().stream(), baggage.values().stream())
                .forEach(contextKey -> {
                    @Nullable final String value = MDC.get(contextKey);
                    if (value == null) return;
                    context.put(contextKey, value);
                });

        return context;
    }
```

Reply:
There should be a dedicated test for baggage as well.

Reply:
isn't baggage added on 
```
    @Override
    public void onBaggage(
            final Tracer tracer,
            final Span span,
            final String baggageKey,
            final String value) 
```
1. Do we need baggage items to be present for logging?
2. Keeping empty map in `ThreadLocal` can be problematic while closing, either we do null check here and add if required or on closing because we need to call `remove`

Reply:
You don't need to call remove